tests/e2e/api/src/models/coupons/coupon.ts

```diff
@@ -0,0 +1,249 @@
+import { Model } from '../model';
+import { HTTPClient } from '../../http';
+import { couponRESTRepository } from '../../repositories';
+import {
+	ModelRepositoryParams,
+	CreatesModels,
+	ListsModels,
+	ReadsModels,
+	UpdatesModels,
+	DeletesModels,
+} from '../../framework';
+import {
+	CouponUpdateParams,
+	CouponDiscountTypeUpdateParams,
+} from './shared';
+
+/**
+ * The parameters that coupons can update.
+ */
+type UpdateParams = CouponUpdateParams & CouponDiscountTypeUpdateParams;
+
+/**
+ * The parameters embedded in this generic can be used in the ModelRepository in order to give
+ * type-safety in an incredibly granular way.
+ */
+export type CouponRepositoryParams = ModelRepositoryParams< Coupon, never, never, UpdateParams >;
+
+/**
+ * An interface for creating coupons using the repository.
+ *
+ * @typedef CreatesCoupons
+ * @alias CreatesModels.<Coupon>
+ */
+export type CreatesCoupons = CreatesModels< CouponRepositoryParams >;
+
+/**
+ * An interface for reading coupons using the repository.
+ *
+ * @typedef ReadsCoupons
+ * @alias ReadsModels.<Coupon>
+ */
+export type ReadsCoupons = ReadsModels< CouponRepositoryParams >;
+
+/**
+ * An interface for updating coupons using the repository.
+ *
+ * @typedef UpdatesCoupons
+ * @alias UpdatesModels.<Coupon>
+ */
+export type UpdatesCoupons = UpdatesModels< CouponRepositoryParams >;
+
+/**
+ * An interface for listing coupons using the repository.
+ *
+ * @typedef ListsCoupons
+ * @alias ListsModels.<Coupon>
+ */
+export type ListsCoupons = ListsModels< CouponRepositoryParams >;
+
+/**
+ * An interface for deleting coupons using the repository.
+ *
+ * @typedef DeletesCoupons
+ * @alias DeletesModels.<Coupons>
+ */
+export type DeletesCoupons = DeletesModels< CouponRepositoryParams >;
+
+/**
+ * The type of discount that is available for the coupon.
+ */
+type DiscountType = 'percent' | 'fixed_cart' | 'fixed_product';
```

Comment:
Can we add `| string` to support custom coupon types?

Reply:
Good idea--done! :)